build/source/driver/multi_driver.f90

```diff
@@ -384,130 +433,139 @@ real(dp),parameter        :: safety = 0.01_dp               ! safety factor to e
-                 ! output: error control
-                 err,message); call handle_err(err,message)
-
- ! calculate vertical distribution of root density
- call rootDensty(mparStruct%hru(iHRU)%var,& ! vector of model parameters
-                 indxStruct%hru(iHRU),    & ! data structure of model indices
-                 progStruct%hru(iHRU),    & ! data structure of model prognostic (state) variables
-                 diagStruct%hru(iHRU),    & ! data structure of model diagnostic variables
-                 err,message)               ! error control
- call handle_err(err,message) 
-
- ! calculate saturated hydraulic conductivity in each soil layer
- call satHydCond(mparStruct%hru(iHRU)%var,& ! vector of model parameters
-                 indxStruct%hru(iHRU),    & ! data structure of model indices
-                 progStruct%hru(iHRU),    & ! data structure of model prognostic (state) variables
-                 fluxStruct%hru(iHRU),    & ! data structure of model fluxes 
-                 err,message)               ! error control
+ ! calculate the fraction of runoff in future time steps
+ call fracFuture(bparStruct%gru(iGRU)%var,    &  ! vector of basin-average model parameters
+                 bvarStruct%gru(iGRU),        &  ! data structure of basin-average variables
+                 err,message)                    ! error control
  call handle_err(err,message)
 
- ! calculate "short-cut" variables such as volumetric heat capacity
- call v_shortcut(mparStruct%hru(iHRU)%var,& ! vector of model parameters
-                 diagStruct%hru(iHRU),    & ! data structure of model diagnostic variables
-                 err,message)               ! error control
- call handle_err(err,message)
-
- ! initialize canopy drip
- ! NOTE: canopy drip from the previous time step is used to compute throughfall for the current time step
- fluxStruct%hru(iHRU)%var(iLookFLUX%scalarCanopyLiqDrainage)%dat(1) = 0._dp 
-
- ! define the file if the first HRU
- if(iHRU==1) then
-  write(fileout,'(a,i0,a,i0,a)') trim(OUTPUT_PATH)//trim(OUTPUT_PREFIX)//'_spinup'//trim(output_fileSuffix)//'.nc'
-  call def_output(nHRU,nSoil(iHRU),fileout,err,message); call handle_err(err,message)
- endif
-
- ! write local model attributes and parameters to the model output file
- call writeAttrb(fileout,iHRU,attrStruct%hru(iHRU)%var,typeStruct%hru(iHRU)%var,err,message); call handle_err(err,message)
- call writeParam(fileout,iHRU,mparStruct%hru(iHRU)%var,bparStruct%var,err,message); call handle_err(err,message)
-
-end do  ! (looping through HRUs)
-
-! allocate space for the upslope area
-allocate(upArea(nHRU),stat=err); call handle_err(err,'problem allocating space for upArea')
+ ! loop through local HRUs
+ do iHRU=1,gru_struc(iGRU)%hruCount
+
+  ! check that the parameters are consistent
+  call paramCheck(mparStruct%gru(iGRU)%hru(iHRU)%var,err,message); call handle_err(err,message)
```

Comment:
At some point it might be nice to put all of this pre-processing into a single subroutine, so that inside these driver loops, we just have a set of essentially 3 calls to `timestep_preprocess` and 'run_timestep` and 'timestep_postprocess`, or something to that effect.